data/crac/crac-impl/src/main/java/com/powsybl/openrao/data/cracimpl/NetworkActionImpl.java

```diff
@@ -7,55 +7,120 @@
+                return Math.abs(danglingLine.getP0() - danglingLineAction.getActivePowerValue().getAsDouble()) >= EPSILON;
+            } else if (elementaryAction instanceof ShuntCompensatorPositionAction shuntCompensatorPositionAction) {
+                ShuntCompensator shuntCompensator = network.getShuntCompensator(shuntCompensatorPositionAction.getShuntCompensatorId());
+                return Math.abs(shuntCompensator.getSectionCount() - shuntCompensatorPositionAction.getSectionCount()) > 0;
+            } else if (elementaryAction instanceof PhaseTapChangerTapPositionAction phaseTapChangerTapPositionAction) {
+                PhaseTapChanger phaseTapChanger = network.getTwoWindingsTransformer(phaseTapChangerTapPositionAction.getTransformerId()).getPhaseTapChanger();
+                return phaseTapChangerTapPositionAction.getTapPosition() != phaseTapChanger.getTapPosition();
+            } else if (elementaryAction instanceof SwitchPair switchPair) {
+                return !network.getSwitch(switchPair.getSwitchToOpen().getId()).isOpen() || network.getSwitch(switchPair.getSwitchToClose().getId()).isOpen();
+            } else if (elementaryAction instanceof TerminalsConnectionAction terminalsConnectionAction) {
+                Identifiable<?> element = network.getIdentifiable(terminalsConnectionAction.getElementId());
+                if (element instanceof Branch<?> branch) {
+                    if (terminalsConnectionAction.isOpen()) {
+                        // Line is considered closed if both terminal are connected
+                        return branch.getTerminal1().isConnected() && branch.getTerminal2().isConnected();
+                    } else {
+                        // Line is already considered opened if one of the terminals is disconnected
+                        return !branch.getTerminal1().isConnected() || !branch.getTerminal2().isConnected();
+                    }
+                } else {
+                    throw new NotImplementedException("TerminalsConnectionAction are only on branches for now");
+                }
+            } else if (elementaryAction instanceof SwitchAction switchAction) {
+                Switch aSwitch = network.getSwitch(switchAction.getSwitchId());
+                return aSwitch.isOpen() != switchAction.isOpen();
+            } else {
+                throw new NotImplementedException();
+            }
+        });
     }
 
     @Override
     public boolean apply(Network network) {
-        if (elementaryActions.stream().anyMatch(elementaryAction -> !elementaryAction.canBeApplied(network))) {
+        if (!canBeApplied(network)) {
             return false;
         } else {
-            elementaryActions.forEach(action -> action.apply(network));
+            elementaryActions.forEach(action -> action.toModification().apply(network, true, ReportNode.NO_OP));
             return true;
         }
     }
 
+    @Override
+    public boolean canBeApplied(Network network) {
```

Comment:
@annetill To check

Reply:
It is iso functional with main, but questionnable.